src/System.Diagnostics.Process/tests/ProcessStreamReadTests.cs

```diff
@@ -88,6 +89,53 @@ public void TestAsyncOutputStream()
             }
         }
 
+        [Fact]
+        // On this test RemotelyInvokable.WriteLoop never returns, we cannot use RemotelyInvokable.WriteLoopUapCmd workaround
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
+        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
+        {
+            var dataArrivedEvent = new AutoResetEvent(false);
+            ConcurrentBag<int> dataReceived = new ConcurrentBag<int>();
+            Process p = CreateProcessPortable(RemotelyInvokable.WriteLoop);
+            try
+            {
+                int minDataArrived = 5;
+                p.StartInfo.RedirectStandardOutput = true;
+                p.OutputDataReceived += (s, e) =>
+                {
+                    if(!string.IsNullOrEmpty(e.Data))
+                    { 
+                        dataReceived.Add(int.Parse(e.Data));
+                    }
+
+                    if(dataReceived.Count > minDataArrived)
+                    { 
+                        dataArrivedEvent.Set();
+                    }
+                };
+                p.Start();
+                p.BeginOutputReadLine();
+            
+                Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Data not come after BeginOutputReadLine");
+                Assert.True(dataReceived.Count >= minDataArrived);
+
+                p.CancelOutputRead();
+
+                // Wait to drain async stream queue
+                Thread.Sleep(3000);
+
+                minDataArrived = 10;
+                p.BeginOutputReadLine();
+
+                Assert.True(dataArrivedEvent.WaitOne(WaitInMS), "Data not come after CancelOutputRead");
+                Assert.True(dataReceived.Count >= minDataArrived);
```

Comment:
I'm not understanding what the test is actually verifying:
- there are race conditions here that could cause this to be true even if it's not behaving as expected
- there's nothing verifying that the callback isn't invoked when it's supposed to not be
- there's nothing verifying that data that was supposed to be thrown away is thrown away
etc.

Reply:
I rewrote test, the only "weak part" I think is 3 seconds sleep to be sure to lose data, I don't find a better way to be sure that queue has drained data, is async and it depends on non deterministic factors, do you have a better idea?

Reply:
I moved callback out of `RemotelyInvokable.cs` because we don't run test on UAP.

Reply:
:sob: named EventWaitHandle is not supported everywhere.
I'll try other solution...without sync primitive it's not so easy write a strong test(with only Thread.Sleep)...do you know on codebase some similar test needs?

Reply:
I could use "shared file+acquire open polling/sleep" to syncronize process on all plat, should be supported everywhere.

Reply:
Previous version with some changes, what do you think?
Final assertion on gaps should ensure:
> there's nothing verifying that the callback isn't invoked when it's supposed to not be
there's nothing verifying that data that was supposed to be thrown away is thrown away

and new `Wait 'WaitInMS' seconds after last signaled data received` should ensure
> there are race conditions here that could cause this to be true even if it's not behaving as expected
```cs
        [OuterLoop]
        [Fact]
        // On this test RemotelyInvokable.WriteLoop never returns, we cannot use RemotelyInvokable.WriteLoopUapCmd workaround
        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)]
        public void TestAsyncOutputStream_BeginCancelBegin_OutputReadLine()
        {
            ConcurrentBag<int> dataReceived = new ConcurrentBag<int>();

            var dataReceivedEvent = new AutoResetEvent(false);

            Process p = CreateProcessPortable(() => 
            {
                int counter = 0;

                for (; ; )
                {
                    Thread.Sleep(500);
                    Console.WriteLine(++counter);
                }
            });

            p.StartInfo.RedirectStandardOutput = true;
            p.OutputDataReceived += (s, e) =>
            {
                if (e.Data != null)
                {
                    Console.WriteLine(e.Data);
                    dataReceived.Add(int.Parse(e.Data));
                    dataReceivedEvent.Set();
                }
            };

            try
            {
                // Start process
                p.Start();

                // Start listening
                p.BeginOutputReadLine();

                // Wait some seconds to accumulate data
                Thread.Sleep(5000);

                // Wait for data signal
                Assert.True(dataReceivedEvent.WaitOne(WaitInMS), "Data not received after BeginOutputReadLine()");

                // Suspend listening
                p.CancelOutputRead();

                // Wait 'WaitInMS' seconds after last signaled data received
                DateTime waitStart = DateTime.UtcNow;
                while (dataReceivedEvent.WaitOne(WaitInMS))
                {
                    Thread.Sleep(1000);

                    // if we'are waiting for more than 'WaitInMS' plus some seconds means that we haven't correctly stopped output
                    Assert.True(DateTime.UtcNow < waitStart.AddMilliseconds(WaitInMS + 5000));
                }

                // We assert that we've received all counter data, with no gaps
                Assert.Empty(Enumerable.Range(dataReceived.OrderBy(v => v).First(), dataReceived.OrderBy(v => v).Last()).Except(dataReceived));

                // Start listening
                p.BeginOutputReadLine();

                // Wait some seconds to accumulate new data
                Thread.Sleep(5000);

                // Wait for data signal
                Assert.True(dataReceivedEvent.WaitOne(WaitInMS), "Data not received after CancelOutputRead()/BeginOutputReadLine()");

                // We assert that we've received new data but we have a gap in serie due to lost number between CancelOutputRead() and BeginOutputReadLine()
                Assert.NotEmpty(Enumerable.Range(dataReceived.OrderBy(v => v).First(), dataReceived.OrderBy(v => v).Last()).Except(dataReceived));
            }
            finally
            {
               p.Kill();
            }
        }

```
test output
```
1
2
3
4
5
6
7
8
9
70
71
72
73
74
75
76
77
78
79
```